### ISSUE TITLE
Validation message isn't used for validation expression

### ISSUE THREAD

**Issue by bjarnef** [umbraco/Umbraco.Forms.Issues#858]
On an Umbraco Cloud project we have the following for an email field type:

![image](https://user-images.githubusercontent.com/2919859/187426462-30cab817-c443-4d47-aaf3-c76586097fa7.png)

The required message is used, but for the validation expression, the message isn't used, but seems to use a standard English message.

![image](https://user-images.githubusercontent.com/2919859/187426727-9185b3d9-c8d4-4095-906e-49e037532b86.png)

![image](https://user-images.githubusercontent.com/2919859/187426778-84618366-2e20-4b98-bfec-5a5ee5e6da7e.png)

**FieldType.Textfield.cshtml**

```
@using Umbraco.Forms.Web
@model Umbraco.Forms.Web.Models.FieldViewModel
@{
    var maxLength = Model.GetSettingValue<int>("MaximumLength", 255);
    var fieldType = Model.GetSettingValue<string>("FieldType", "text");
    var autocompleteAttribute = Model.GetSettingValue<string>("AutocompleteAttribute", string.Empty);
}

<input type="@fieldType" name="@Model.Name" id="@Model.Id" data-umb="@Model.Id" class="@Html.GetFormFieldClass(Model.FieldTypeName)" value="@Model.ValueAsHtmlString" maxlength="@maxLength"
       @{if (string.IsNullOrEmpty(Model.PlaceholderText) == false) { <text> placeholder="@Model.PlaceholderText" </text> }}
       @{if (string.IsNullOrEmpty(autocompleteAttribute) == false) { <text> autocomplete="@autocompleteAttribute" </text> }}
       @{if (Model.Mandatory || Model.Validate) { <text> data-val="true" </text> }}
       @{if (Model.Mandatory) { <text> data-val-required="@Model.RequiredErrorMessage" </text> }}
       @{if (Model.Validate) { <text> data-val-regex="@Model.InvalidErrorMessage" data-val-regex-pattern="@Html.Raw(Model.Regex)" </text> }} />
```

Umbraco v10.0.1
Forms v10.0.5

---
_This item has been added to our backlog AB#69867_

**Comment by bjarnef** [comment]
@AndyButland are you able to reproduce this?

We are now on:

Umbraco v10.2.1
Forms v10.1.3

but the error message for validation expression isn't respected:

![image](https://user-images.githubusercontent.com/2919859/196679204-ffc65021-98b4-4561-97f0-bd15feb18909.png)

![image](https://user-images.githubusercontent.com/2919859/196678733-faa6a22c-085e-435e-bdef-dc28b15ab8d9.png)

**Comment by TRexStark** [comment]
This is still an issue on Forms 10.2.3

**Comment by bjarnef** [comment]
@AndyButland can you check if this has been fixed in recent changes or if it still an issue?

**Comment by AndyButland** [comment]
No, it's still an issue.  If you know the background to it it's actually kind of expected, but I can see for someone coming across it it appears to be a bug.

To summarise - the validation we've had for a long time with Forms is the one via the regular expression, where you can configure a pattern to validate against and a message to display if it fails.  There isn't any issue with this to my knowledge.

In a more recent version of Forms we added the option to set the "field type" for a text field, of which one of the options is "email".  If you enter an invalid email, the browser will get involved directly and display the message you see.  But it doesn't know anything about the validation message you've configured here, as that isn't invoked.

Ideally we'll find a way in this situation, such that the browser message can be initialised from the validation expression message, if that's set to "email" and has been set.

**Comment by bjarnef** [comment]
@AndyButland okay, however from a UI/UX perspective (or content editors manage the form), it seems odd the mandatory message is displayed, but not the message at regular validation expression.

![image](https://user-images.githubusercontent.com/2919859/224666219-2562e5cc-566a-4d55-9f95-c0c780874ef8.png)

Not sure if this is used in other contexts though?

**Comment by AndyButland** [comment]
I had a further look into this one today.  Just to recap, the problem is that with Umbraco Forms you can both set a validation regular expression/message AND you can define the type of the input field as "email".  In that situation the browser's default message for an invalid email takes precedence over the one defined in Umbraco Forms.

What we ideally want to do is to override the browser's message with the Umbraco Forms one.

I managed to get this working, which you should be able to try out if you create a file at `Views\Partials\Forms\Themes\default\Fieldtypes\FieldType.Textfield.cshtml` and copy in the following code.

```
@using Umbraco.Forms.Web
@model Umbraco.Forms.Web.Models.FieldViewModel
@{
    var maxLength = Model.GetSettingValue<int>("MaximumLength", 255);
    var fieldType = Model.GetSettingValue<string>("FieldType", "text");
    var autocompleteAttribute = Model.GetSettingValue<string>("AutocompleteAttribute", string.Empty);
    var hasCustomInvalidEmailMessage = string.Equals(fieldType, "email", StringComparison.InvariantCultureIgnoreCase) &&
        Model.Validate &&
        !string.IsNullOrWhiteSpace(Model.InvalidErrorMessage);
}
<input style="background-color: palegreen" type="@fieldType" name="@Model.Name" id="@Model.Id" data-umb="@Model.Id" class="text @Html.GetFormFieldClass(Model.FieldTypeName)" value="@Model.ValueAsHtmlString" maxlength="@maxLength"
       @{if (string.IsNullOrEmpty(Model.PlaceholderText) == false) { <text> placeholder="@Model.PlaceholderText" </text> }}
       @{if (string.IsNullOrEmpty(autocompleteAttribute) == false) { <text> autocomplete="@autocompleteAttribute" </text> }}
       @{if (Model.Mandatory || Model.Validate) { <text> data-val="true" </text> }}
       @{if (Model.Mandatory) { <text> data-val-required="@Model.RequiredErrorMessage" aria-required="true" </text> }}
       @{if (Model.Validate) { <text> data-val-regex="@Model.InvalidErrorMessage" data-val-regex-pattern="@Html.Raw(Model.Regex)" </text> }}
       @{if (hasCustomInvalidEmailMessage) { <text> title="@Model.InvalidErrorMessage" oninvalid="this.setCustomValidity('@Model.InvalidErrorMessage.Replace("'", "\'")');" oninput="setCustomValidity('')" </text> }}
       @{if (!string.IsNullOrEmpty(Model.ToolTip)) { <text> aria-describedby="@(Model.Id)_description" </text> } }/>
```

If you then open up a form on the front-end that uses a Short Answer field, defined with a custom email validation and a field type of "email", you'll see the updated behaviour.  I made it green just so you can be sure you are seeing the custom partial view and not the built-in one.  Otherwise the lines related to the `hasCustomInvalidEmailMessage` flag are the only changes.

It works, but only if you set the `title` attribute.  Which is fine if the message is something like "Please enter a valid email address." but not so much if the message is along the lines of "Sorry, that isn't a valid email".  As you see this message when you hover over the input even if there's no entry in it yet.

So not sure yet if this is a good idea to add to Forms as the default behaviour for the field, but wondered what your thoughts were @bjarnef or if you can see if we could improve it.  Thanks.

**Comment by rickbutterfield** [comment]
Hey all!

A fix for this is now in Forms 17.5.0-rc / 18.1.0-rc, released today. Please give it a try and report back if you can!

**Comment by rickbutterfield** [comment]
The fix has now landed in the final 17.5.0/18.1.0 release
